Add support for custom properties

jiayihu opened the pull request:
This is an attempt to start improving the support for CSSNext: https://github.com/atom/language-css/issues/51
It's also my first time with a language package so please tell me if I did something wrong.

50Wliu commented on grammars/css.cson:
```diff
@@ -483,6 +483,10 @@
         'include': '#comment-block'
       }
       {
+        'match': '--([A-Za-z0-9_-]+)'
```
This needs a more complex pattern in order to avoid matching `------` as a variable, which I'm pretty sure is invalid.  @esdoppio can probably give some tips regarding this.

jiayihu replied:
I improved it with `--([A-Za-z][A-Za-z0-9_-]*)`, but I'm open to suggestions.

esdoppio replied:
"------" is indeed a really lousy name, but it's valid (test it yourself), so I think we should allow it (to be tokenized) and let people make their own decisions.

50Wliu replied:
Huh.  I stand corrected.

50Wliu commented on grammars/css.cson:
```diff
@@ -483,6 +483,10 @@
       }
       {
+        'match': '--([A-Za-z0-9_-]+)'
+        'name': 'variable.scss'
```
This is language-css, so please use `variable.css`.

jiayihu replied:
I replaced all `scss`occurrences in the grammar to `css`

50Wliu replied:
Wow, when did the others get there? 🙈

50Wliu commented:
* `var(--variable-name)` support needs to be added.
* Is the `:root` pseudo-selector matched?
* Specs are required.

jiayihu commented:
- var(--variable-name) support was [already added](https://github.com/jiayihu/language-css/blob/56e6f113de251db83b31cb6c5e9d0c8e1e97e96b/grammars/css.cson#L349) by someone else
- The `:root` pseudo selector is matched as any other pseudo-selector. Its scope is indeed `source.css, meta.selector.css, entity.other.attribute-name.pseudo-class.css`
- I added the specs. First time writing in CS 😅  

I noticed that the specs for `var(--variable-name)` are probably missing, should I add them also? Also the pattern for `var(--variable)` matches `var(------)` too as my previous RegExp, so I believe it should be fixed too.

50Wliu commented:
> I noticed that the specs for var(--variable-name) are probably missing, should I add them also? Also the pattern for var(--variable) matches var(------) too as my previous RegExp, so I believe it should be fixed too.

If you could do both of these, that would be ✨

esdoppio commented on grammars/css.cson:
```diff
@@ -483,7 +483,8 @@
       }
       {
-        'match': '--([A-Za-z0-9_-]+)'
+        'match': '--([A-Za-z][A-Za-z0-9_-]*)'
```
The pattern is a bit too strict.

From https://www.w3.org/TR/css-variables/#custom-property:

> it’s defined as any valid identifier that starts with two dashes

And a *valid* identifier shares the same syntax as an `<ident-token>`—(escaping set aside) it

  * **may contain only**
    * **[a-zA-Z0-9_-]**
    * **ISO 10646 characters U+00A0 and higher**
  * must not begin with \[0-9]
  * must not begin with '-' followed by \[0-9]
  * must be at least 2 characters long if begins with '-'

Since the identifier in this case always starts with `--`, the last three requirements are fulfilled, you only have to care about the first rule (in bold) here. So a simple

```
--[\\w-]+
```

suffices. Just in case, `\w` is not equivalent to `[a-zA-Z0-9_]` with the Unicode defined encodings like UTF-8. 😉

esdoppio replied:
Actually, even

```
--[\\w-]+
```

is too strict regarding the syntax, but since no one is really complaining about they want to use emoji or escaping in, like, class names, I think we're good.

jiayihu commented:
Okay so...
- [x] Updated the Regexp for both custom properties and custom variables
- [x] Added spec for both
- [x] Squashed some commits in order to have a cleaner history

Anything else that I missed? 🤔

50Wliu commented:
Thanks!
